## 실습: 타이타닉 데이터 분석 (Pandas)
----
* Pandas에서 배운 내용을 **실제 데이터셋**에 적용해보는 실습
* 데이터셋: **타이타닉(Titanic)** 승객 데이터 (891명)
    - 1912년 타이타닉호 침몰 사고 승객들의 생존 여부와 정보
    - 데이터 분석 입문에서 가장 많이 사용되는 대표 데이터셋
* 진행 순서
    1. 데이터 불러오기
    2. 데이터 살펴보기 (Pandas 조회)

In [1]:
## 패키지 불러오기
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns   # 데이터셋을 불러오기 위해서만 사용

### 1) 데이터 불러오기

* ```seaborn``` 패키지에 내장된 타이타닉 데이터셋을 ```pandas.DataFrame```으로 불러옴
* 각 열(column)의 의미

| 열 이름 | 의미 |
|---|---|
| survived | 생존 여부 (0 = 사망, 1 = 생존) |
| pclass | 객실 등급 (1, 2, 3등석) |
| sex | 성별 |
| age | 나이 |
| sibsp | 함께 탑승한 형제/배우자 수 |
| parch | 함께 탑승한 부모/자녀 수 |
| fare | 운임(요금) |
| embarked | 탑승 항구 (C = Cherbourg, Q = Queenstown, S = Southampton) |
| class | 객실 등급 (문자) |
| who | man / woman / child |
| adult_male | 성인 남성 여부 |
| deck | 객실 데크 (A~G) |
| embark_town | 탑승 도시 |
| alive | 생존 여부 (문자) |
| alone | 혼자 탑승 여부 |

In [2]:
df = sns.load_dataset('titanic')
print(type(df))

<class 'pandas.core.frame.DataFrame'>


### 2) 데이터 살펴보기

* 분석을 시작하기 전에 데이터가 어떤 모양인지 확인하는 단계
* ```head()```, ```tail()```, ```shape```, ```info()```, ```describe()``` 등을 사용

In [ ]:
# 앞부분 5개 행 조회
df.head()

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True


In [ ]:
# 뒷부분 3개 행 조회
df.tail(3)

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
888,0,3,female,NaN,1,2,23.45,S,Third,woman,False,NaN,Southampton,no,False
889,1,1,male,26.0,0,0,30.00,C,First,man,True,C,Cherbourg,yes,True
890,0,3,male,32.0,0,0,7.75,Q,Third,man,True,NaN,Queenstown,no,True


In [ ]:
# 크기 (행, 열) -> numpy의 shape와 동일한 개념
print('shape :', df.shape)
print('행의 수 :', len(df))
print('열의 수 :', len(df.columns))

shape : (891, 15)
행의 수 : 891
열의 수 : 15


In [ ]:
# 열 이름, 인덱스, 값
print(df.columns) # 열 이름
print(df.index) 
print(type(df.values), df.values.shape)   # 값 부분은 numpy 배열

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')
RangeIndex(start=0, stop=891, step=1)
<class 'numpy.ndarray'> (891, 15)


In [7]:
# 각 열의 자료형과 비어 있지 않은 값(Non-Null)의 개수
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.7+ KB


* ```info()``` 결과를 보면 전체 891행인데 ```age```는 714개, ```deck```은 203개만 값이 있음
    - 나머지는 값이 비어 있는 결측치(```NaN```)
    - 결측치를 어떻게 처리하는지는 **3-3 데이터 전처리**에서 배움
    - 이 실습에서는 나이 계산이 필요할 때 ```dropna()```로 비어 있는 값을 **제외하고** 사용

In [ ]:
# 나이 열에서 비어 있는 값을 제외한 Series
age = df['age'].dropna()
print('전체 행 수 :', len(df['age']))
print('나이가 있는 행 수 :', len(age))

전체 행 수 : 891
나이가 있는 행 수 : 714


In [9]:
# 숫자형 열의 기초 통계량 (개수, 평균, 표준편차, 최소, 사분위수, 최대)
df.describe()

,survived,pclass,age,sibsp,parch,fare
count,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


#### 열(column) 선택하기
* ```df['열이름']``` : 하나의 열 -> ```Series```
* ```df[['열이름1', '열이름2']]``` : 여러 열 -> ```DataFrame```

In [10]:
# 하나의 열 선택 -> Series
fares = df['fare']
print(type(fares))
print(fares.head())

<class 'pandas.core.series.Series'>
0     7.2500
1    71.2833
2     7.9250
3    53.1000
4     8.0500
Name: fare, dtype: float64


In [11]:
# 여러 열 선택 -> DataFrame
sub = df[['survived', 'pclass', 'sex', 'age', 'fare']]
print(type(sub))
sub.head()

<class 'pandas.core.frame.DataFrame'>


,survived,pclass,sex,age,fare
0,0,3,male,22.0,7.2500
1,1,1,female,38.0,71.2833
2,1,3,female,26.0,7.9250
3,1,1,female,35.0,53.1000
4,0,3,male,35.0,8.0500


#### 행(row) 선택하기
* ```df.loc[인덱스]``` : 인덱스 **이름**으로 선택
* ```df.iloc[번호]``` : **순서(정수 위치)**로 선택 -> numpy 인덱싱/슬라이싱과 같은 방식

In [12]:
# 0번 행 하나 선택
print(df.loc[0])

survived                 0
pclass                   3
sex                   male
age                   22.0
sibsp                    1
parch                    0
fare                  7.25
embarked                 S
class                Third
who                    man
adult_male            True
deck                   NaN
embark_town    Southampton
alive                   no
alone                False
Name: 0, dtype: object


In [13]:
# iloc: numpy처럼 [행 슬라이스, 열 슬라이스]
print(df.iloc[0:3, 0:4])   # 0~2행, 0~3열

   survived  pclass     sex   age
0         0       3    male  22.0
1         1       1  female  38.0
2         1       3  female  26.0


In [14]:
# loc: 이름으로 행/열 지정
df.loc[0:4, ['sex', 'age', 'survived']]

,sex,age,survived
0,male,22.0,0
1,female,38.0,1
2,female,26.0,1
3,female,35.0,1
4,male,35.0,0


#### 조건으로 행 골라내기 (불리언 인덱싱)
* numpy에서 ```arr[arr > 0]``` 처럼, 조건식의 결과(True/False)를 이용해 행을 선택
* 여러 조건은 ```&``` (and), ```|``` (or) 로 연결하고 각 조건을 괄호로 묶음

In [15]:
# 조건식 자체는 True/False로 이루어진 Series
cond = df['age'] < 10
print(cond.head())
print('10세 미만 승객 수 :', cond.sum())   # True를 1로 계산

0    False
1    False
2    False
3    False
4    False
Name: age, dtype: bool
10세 미만 승객 수 : 62


In [16]:
# 조건으로 행 선택
kids = df[df['age'] < 10]
kids.head()

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
7,0,3,male,2.0,3,1,21.0750,S,Third,child,False,NaN,Southampton,no,False
10,1,3,female,4.0,1,1,16.7000,S,Third,child,False,G,Southampton,yes,False
16,0,3,male,2.0,4,1,29.1250,Q,Third,child,False,NaN,Queenstown,no,False
24,0,3,female,8.0,3,1,21.0750,S,Third,child,False,NaN,Southampton,no,False
43,1,2,female,3.0,1,2,41.5792,C,Second,child,False,NaN,Cherbourg,yes,False


In [17]:
# 여러 조건: 1등석 여성 승객
first_female = df[(df['pclass'] == 1) & (df['sex'] == 'female')]
print('1등석 여성 승객 수 :', len(first_female))
first_female.head()

1등석 여성 승객 수 : 94


,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
11,1,1,female,58.0,0,0,26.5500,S,First,woman,False,C,Southampton,yes,True
31,1,1,female,NaN,1,0,146.5208,C,First,woman,False,B,Cherbourg,yes,False
52,1,1,female,49.0,1,0,76.7292,C,First,woman,False,D,Cherbourg,yes,False


In [18]:
# 범주형 값의 개수 세기
print(df['pclass'].value_counts())
print()
print(df['embark_town'].value_counts())

pclass
3    491
1    216
2    184
Name: count, dtype: int64

embark_town
Southampton    644
Cherbourg      168
Queenstown      77
Name: count, dtype: int64


In [19]:
# 고유한 값 확인
print(df['sex'].unique())
print(df['who'].unique())
print(df['embark_town'].unique())

['male' 'female']
['man' 'woman' 'child']
['Southampton' 'Cherbourg' 'Queenstown' nan]


In [20]:
## [연습 2-1] 3등석(pclass == 3) 승객 중 나이가 60세 이상인 승객을 골라 sex, age, fare 열만 출력하세요.
#####################################
# Your code here
#####################################


## [연습 2-2] 운임(fare)이 100 이상인 승객은 몇 명인지 출력하세요.
#####################################
# Your code here
#####################################


## [연습 2-3] iloc을 사용해 마지막 5개 행의 앞 3개 열을 출력하세요.
#####################################
# Your code here
#####################################